### ISSUE TITLE
Support CSV Output from `pidstat`

### ISSUE THREAD

**Issue by tst-ppenev** [sysstat/sysstat#174]
As a `pidstat` user, I would like to be able to receive the statistics for each time period as a CSV file.  That makes it easier to graph results, and correlate them with data from profiling tools.

Note from issue creator:

I'm not familiar with conventions for contributing to the SysStat project, and I didn't see a developer guide, so this is my best effort to open a ticket and submit a pull request.  Advice is appreciated.

**Comment by sysstat** [comment]
Thank you for your contribution.
First, I have to tell you that I been unable to pull any changes from your development branch:
```
$ git checkout -b tst-ppenev-#174-pidstat-csv-output master
Switched to a new branch 'tst-ppenev-#174-pidstat-csv-output'

$ git pull https://github.com/tst-ppenev/sysstat.git #174-pidstat-csv-output
From https://github.com/tst-ppenev/sysstat
 * branch            HEAD       -> FETCH_HEAD
Already up-to-date.
```
I didn't have time to google for an explanation. Anyway I mean that I may have other remarks once I know how to fix the problem. Some remarks below may even don't apply if I didn't understand properly how your patch works.

So here are several things that you should take into account before this feature can get included in sysstat/pidstat. The idea is to keep consistent with how other options from other sysstat commands work:

* Don’t create a CSV file in addition to the report already displayed on standard output. If the user selects the CSV output format then pidstat will display its statistics only in this format to the standard output. To create a CSV file, the user will need to redirect the output to a file. See how things are done for other sysstat commands that already support a different output format (e.g. iostat and mpstat with their JSON output format).
* This is not pidstat’s goal to give the user access to the kernel’s counters value. Pidstat calculates metrics for tasks (number of I/O per second, CPU utilization, etc.) and display them. The fact that the output will be a CSV format should not change this. Display the metrics, not the counters value (or the number of clock ticks...)
* Keep consistent with the option’s name. iostat and mpstat use “-o JSON” to output their data in JSON format. I would like you to use “-o CSV” here in pidstat.
* I know that CSV actually means Comma-Separated Values, but please use a semi-colon instead of a comma as a delimiter since the comma may be used as the decimal point in certain countries.
* Please also comply with sysstat’s coding standards. In particular your comments should look like this:

`/* This is a comment on a single line */`
Or
```
/*
 * This is a comment written on
 * several lines.
 */
```
But not:
`// This is your comment:`

**Comment by tst-ppenev** [comment]
Hi, Sebastien.

Thanks for the comments.  I can address some of them.  However, my team is trying to use `pidstat` for profiling running executables.  Having detailed information like clock tick counts for each thread can be very useful for that purpose.  However, if that is not a use case for `pidstat`, we can fork off the collection of that type of data into a separate tool.

Here's what I have been able to use to check out my feature branch.  (I believe your `git checkout` command is wrong.  It creates a new branch called `tst-ppenev-#174-pidstat-csv-output master`, rather than checking out an existing branch called `#174-pidstat-csv-output`.)

```BASH
git clone https://github.com/tst-ppenev/sysstat.git
cd sysstat
git checkout '#174-pidstat-csv-output'
```

Q: Are you able to comment on the individual code lines of the pull request at https://github.com/sysstat/sysstat/pull/175/files?

**Comment by tst-ppenev** [comment]
@sysstat: The reason I added an output file path to the CSV format switch is that I didn't find an easy way to separate the output of `pidstat`.  How do you separate the standard output of `pidstat` from the standard output of the command you're getting statistics for.

E.g., if you create

`/tmp/sample-command.sh`:
```BASH
! /bin/bash

echo "$0:" "The output is standard."
sleep 2
echo "$0:" "The standard is output."
```

and run:

```BASH
pidstat 1 -e /tmp/sample-command.sh > sample-command.pidstat-output.txt
```

You get:

`sample-command.pidstat-output.txt`:
```
Linux 4.13.0-36-generic (pav-ThinkPad-P51s) 	03/05/2018 	_x86_64_	(4 CPU)
/tmp/sample-command.sh: The output is standard.

05:56:24 PM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
05:56:25 PM  1000     17304    0.00    0.00    0.00    0.00    0.00     3  sample-command.
05:56:26 PM  1000     17304    0.00    0.00    0.00    0.00    0.00     3  sample-command.
/tmp/sample-command.sh: The standard is output.

Average:     1000     17304    0.00    0.00    0.00    0.00    0.00     -  sample-command.
```

**Comment by sysstat** [comment]
Yes, that's why pidstat should display either its standard report *or* its CSV output, not both.
I hope to get some time to test the latest version of your patch at the end of this week. I will then get back in touch with you.

**Comment by reikred** [comment]
@tst-ppenev @sysstat Hi, it sounds good to have CSV output option for pidstat, but how about getting an option also for binary output so that I can postsprocess and make graphs with sadf? 

My version is 11.3.5-1.fc25

**Comment by franzwong** [comment]
@sysstat Is there any chance to add csv support?

**Comment by franzwong** [comment]
Anyhow, I've added csv format and no header options. But it is not quite polished (with lots of if-else condition for csv display), so I don't create a pull request.

https://github.com/franzwong/sysstat/tree/support-csv

**Comment by sysstat** [comment]
> @sysstat Is there any chance to add csv support?

Yes, this could happen, but sorry I'm really, really too much busy at the present time...

**Comment by aparcar** [comment]
I'd still be keen to have this feature, would it be possible to add it somehow?

**Comment by sysstat** [comment]
> I'd still be keen to have this feature, would it be possible to add it somehow?

In fact I had added support for JSON output for some sysstat commands like `iostat` and `mpstat` (and `sar` of course) at the beginning. To be consistent, I have also added support for JSON output for all other sysstat commands, including `pidstat`. CSV output appears to be more questionnable now.

**Comment by aparcar** [comment]
Oh wonderful! Since the latest version of `pidstat` isn't part of my used distribution yet, I didn't notice that.

Whoever is in the same situation, I ended up creating a Python script to convert the values:

<details>

```python
#!/usr/bin/python3 -u

import sys

data = {}

print("time,pid,cpu_perc,vsz,rss,mem_perc")
now_cpu = False
now_mem = False
for line in sys.stdin:
    columns = line.split()

    if not columns or len(columns) < 2:
        continue

    if columns[0] == "Average":
        break

    # Parse the time, PID, and different sections
    if columns[3] == '%usr' and 'CPU' in line:  # CPU usage line
        now_cpu = True
        continue

    if now_cpu:
        data["time"] = columns[0]
        data["pid"] = columns[2]
        data["cpu_perc"] = columns[7]
        now_cpu = False

    if columns[3] == 'minflt/s':
        now_mem = True
        continue

    if now_mem:
        data["VSZ"] = columns[5]
        data["rss"] = columns[6]
        data["mem_perc"] = columns[7]
        now_mem = False

    elif columns[3] == "minflt-nr":
        print(f"{data['time']},{data['pid']},{data['cpu_perc']},{data['VSZ']},{data['rss']},{data['mem_perc']}")
        data.clear()
```

</details>

And run it like this:

```bash
exec pidstat -H -T ALL -p "$TEST_PID" 1 -r -u | pidstat2csv.py > $TEST_OUT
```

Result:

<details>

```csv
time,pid,cpu_perc,vsz,rss,mem_perc
1742214528,939,2.00,14644,8880,0.11
1742214529,939,0.00,14644,8880,0.11
1742214530,939,0.00,14644,8880,0.11
1742214531,939,0.00,14644,8880,0.11
1742214532,939,0.00,14644,8880,0.11
1742214533,939,0.00,14644,8880,0.11
1742214534,939,0.00,14644,8880,0.11
1742214535,939,0.00,14644,8880,0.11
1742214536,939,0.00,14644,8880,0.11
1742214537,939,0.00,14644,8880,0.11
1742214538,939,0.00,14644,8880,0.11
1742214539,939,0.00,14644,8880,0.11
1742214540,939,0.00,14644,8880,0.11
1742214541,939,0.00,14644,8880,0.11
```
</details>
